### PR TITLE
[wallet] fee fixes: always create change, adjust value, and p…

### PR DESCRIPTION
**Pull request by instagibbs**
…rune later

This PR is an attempt to solve some of https://github.com/bitcoin/bitcoin/issues/10247 , first and foremost the case:

> // TODO: The case where there is no change output remains
>// to be addressed so we avoid creating too small an output

This is resolved by always assuming a change output exists, and rebalancing that value by the fee excess, and deleting when necessary. The logic is also simpler to me.

### CODE REVIEW

**Comment by instagibbs** [comment]
Note: This is a "regression" due to `MIN_FINAL_CHANGE` being taken out of the calculation: https://github.com/bitcoin/bitcoin/pull/10333/files#diff-b2bb174788c7409b671c46ccc86034bdL2647

A bit confusing since the change is dropped when `IsDust` as before... this would be a good time to sync this behavior.

**Comment by instagibbs** [comment]
rebased, and included logic to attempt change larger than `MIN_FINAL_CHANGE` when possible. In the end some cases cannot be avoided, and just-over-dust change may be kept.

**Comment by instagibbs** [comment]
suggested new strategy: instead of "trying one more time" when change is in middle-ground, cache the current transaction, increase the amount you're attempting to send, and try again. If you then run out of available coins, you simply return the cached version of the transaction.

**Comment by instagibbs** [comment]
Implemented the previously mentioned idea. Instead of doing "one more try" to get acceptable change, it caches the successful yet small change transaction and slowly grows the amount of coins it grabs until it creates an acceptable transaction, given reasonable chosen inputs this will result in a transaction with change larger than MIN_FINAL_CHANGE. If it fails and runs out of coins it returns the cached transaction.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2709,41 +2712,54 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
...
+                    else if (IsDust(txNew.vout[nChangePosInOut], ::dustRelayFee)) {
+                        nFeeRet += txNew.vout[nChangePosInOut].nValue;
+                        txNew.vout.erase(txNew.vout.begin() + nChangePosInOut);
+                        nChangePosInOut = -1;
+                        reservekey.ReturnKey();
+                    // If larger than dust, but still small, increase fee target and try again
+                    } else if (txNew.vout[nChangePosInOut].nValue < MIN_FINAL_CHANGE) {
+                        // Save this transaction, use if we cannot get large-enough change
+                        if (!have_cached_txn) {
+                            tx_cached = txNew;
```
Comment (achow101) [inline src/wallet/wallet.cpp]: `have_cached_txn` should probably be set to true here as it is never actually set anywhere

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2501,6 +2501,13 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
     wtxNew.fTimeReceivedIsTxTime = true;
     wtxNew.BindWallet(this);
     CMutableTransaction txNew;
+    // This transaction is used to track the best transaction that has change
+    // falling between uneconomical dust and MIN_FINAL_CHANGE. If the wallet
+    // runs out of funds trying to find a transaction that has pruneable
+    // change dust or change larger than MIN_FINAL_CHANGE, it will
+    // return this transaction.
+    CMutableTransaction tx_cached;
+    bool have_cached_txn;
```
Comment (achow101) [inline src/wallet/wallet.cpp]: Shouldn't you assign `have_cached_txn` to false here as it is not assiged before it is used down below?

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2592,73 +2599,69 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
...
-                        std::vector<CTxOut>::iterator position = txNew.vout.begin()+nChangePosInOut;
-                        txNew.vout.insert(position, newTxOut);
-                    }
+                // TODO move this section near fee-adjustment area.
+                if (nChangePosInOut == -1)
+                {
+                    // Insert change txn at random position:
+                    nChangePosInOut = GetRandInt(txNew.vout.size()+1);
+                }
+                else if ((unsigned int)nChangePosInOut > txNew.vout.size())
```
Comment (achow101) [inline src/wallet/wallet.cpp]: I don't see how this case matters. Wouldn't `nChangePosInOut` never be set to an out of range index?

Comment (instagibbs) [inline src/wallet/wallet.cpp]: Caller may request a too-high change position. This is also a simple code move, not going to change this logic.

**Comment by instagibbs** [comment]
fixed issues, thanks @achow101

**Review by gmaxwell — COMMENTED** [review]
Concept ACK

**Comment by morcos** [comment]
I think #10712 is preferable to this PR

**Comment by instagibbs** [comment]
closing in favor of #10712

**Pull request closed without merging**